tree-walk-xl.go

```diff
@@ -91,14 +82,15 @@ func (xl xlObjects) treeWalk(bucket, prefixDir, entryPrefixMatch, marker string,
 		}
 	}
 	entries, err := xl.listDir(bucket, prefixDir, func(entry string) bool {
-		return !strings.HasPrefix(entry, entryPrefixMatch)
+		return strings.HasPrefix(entry, entryPrefixMatch)
 	}, isLeaf)
 	if err != nil {
-		send(treeWalkResult{err: err})
-		return false
+		ch <- treeWalker{err: err}
+		return
 	}
 	if len(entries) == 0 {
-		return true
+		ch <- treeWalker{end: true}
```

Comment:
this is not correct, because if one of the subdirs has 0 entries it will send `end: true` so the upper directories will not be listed, also if the channel is closed it will get panic.